packages/babel-plugin-transform-simplify-comparison-operators/src/index.js

```diff
@@ -1,6 +1,67 @@
 "use strict";
 
-module.exports = function() {
+module.exports = function({ types: t }) {
+  // custom implementation of getTypeAnnotation that fixes
+  // the type information that is lost during sequence expression transformation
+  // https://github.com/babel/minify/issues/323
+  function customTypeAnnotation(path) {
+    if (path.typeAnnotation) {
+      return path.typeAnnotation;
+    }
+
+    // We are not handling the case of literals and other base types
+    // since they are already handled via getTypeAnnotation
+    const { node } = path;
+    const binding = path.parentPath.scope.getBinding(node.name);
+
+    const types = [];
+    if (binding && binding.constantViolations) {
+      if (binding.identifier.typeAnnotation) {
+        return binding.identifier.typeAnnotation;
+      }
+      if (binding.constantViolations) {
+        const violations = binding.constantViolations;
+        for (let violation of violations) {
+          types.push(violation.getTypeAnnotation());
+        }
+      }
+    }
+
+    if (types.length > 0) {
+      return t.createUnionTypeAnnotation(types);
+    }
+    return types;
+  }
+
+  // Based on the type inference in Babel
+  function baseTypeStrictlyMatches(left, right) {
+    let leftTypes, rightTypes;
+
+    if (t.isIdentifier(left)) {
+      leftTypes = customTypeAnnotation(left);
+    } else if (t.isIdentifier(right)) {
+      rightTypes = customTypeAnnotation(right);
+    }
+
+    // Early exit
+    if (t.isAnyTypeAnnotation(leftTypes) || t.isAnyTypeAnnotation(rightTypes)) {
+      return false;
+    }
+
+    leftTypes = [].concat(leftTypes, left.getTypeAnnotation());
+    rightTypes = [].concat(rightTypes, right.getTypeAnnotation());
+
+    leftTypes = t.createUnionTypeAnnotation(leftTypes);
+    rightTypes = t.createUnionTypeAnnotation(rightTypes);
+
+    if (
+      !t.isAnyTypeAnnotation(leftTypes) &&
+      t.isFlowBaseAnnotation(leftTypes)
```

Comment:
I'm not sure if this affects the babel-7 upgrade. I think we should test this against Babel7. What do you think?

Reply:
No it stays the same, No change for babel 7 

https://github.com/babel/babel/blob/64dfb7574e924be3ad737c33cbc81d619dbc14c6/packages/babel-traverse/src/path/inference/index.js#L109-L111